libyara/grammar.y

```diff
@@ -1538,13 +1641,26 @@ string_enumeration_item
 
 for_expression
     : primary_expression
+      {
+        int mem_offset = LOOP_LOCAL_VARS * compiler->loop_depth;
+        yr_parser_emit_with_arg(yyscanner, OP_CLEAR_M, mem_offset + 4, NULL, NULL);
+        yr_parser_emit_with_arg(yyscanner, OP_ADD_M, mem_offset + 4, NULL, NULL);
+        yr_parser_emit_with_arg(yyscanner, OP_PUSH_M, mem_offset + 4, NULL, NULL);
```

Comment:
Here we have a push that didn't exist before and it should be pulled somewhere, however `for_expression` is used in loops of the form `for_expression _OF_ string_set` for which nothing else has changed. What happens with that extra push in those loops? 

The test cases run ok, but maybe it's because we are not covering that case.

Reply:
I didn't noticed that `OP_ADD_M` pops an item from the stack, so the `OP_PUSH_M` is just to put it in the stack again. So this code doesn't leave an item in the stack as I thought at first. However, it adds unnecessary extra instructions when `for_expression` is used in `for_expression _OF_ string_set` and `for for_expression _OF_ string_set : ` statements. For example:

`4 of ($a, $b)` 

results in ...
```
01.   OP_INIT_RULE
02.   OP_PUSH 4      
03.   OP_CLEAR_M 4
04.   OP_ADD_M 4
05.   OP_PUSH_M 4     
05.   OP_PUSH UNDEF 
06.   OP_PUSH $a   
07.   OP_PUSH $b
08.   OP_OF 
09.   OP_MATCH_RULE
```

Instructions 3, 4, and 5 are not required at all in this case.

My proposal is leaving the `for_expression` production as:

```
for_expression
    : primary_expression
      {
        $$ = FOR_EXPRESSION_ANY;
      }
    | _ALL_
      {
        $$ = FOR_EXPRESSION_ALL;
      }
    | _ANY_
      {
        $$ = FOR_EXPRESSION_ANY;
      }
    ;
```

And emit the required instructions in `_FOR_ for_expression _IDENTIFIER_ _IN_`.